Message: parsing time "2024-10-16T14:36:58 02:00" as "2006-01-02T15:04:05Z07:00": cannot parse " 02:00" as "Z07:00".

### Describe the bug

Here's an example on using `.sinceTime()`

https://github.com/fabric8io/kubernetes-client/blob/08e54a9fa6da1d7eacd894f8251f3ae7848b6961/doc/CHEATSHEET.md#L2342-L2345

When I use this with a timezoned timestamp with fabric8 kubernetes client 6.13.0, I get an exception from the remote:

```
Failure executing: GET at: https://api.crc.testing:6443/api/v1/namespaces/redhat-ods-operator/pods/rhods-operator-76bb96fb9d-qg2cs/log?pretty=false&sinceTime=2024-10-16T14:36:58+02:00. Message: parsing time "2024-10-16T14:36:58 02:00" as "2006-01-02T15:04:05Z07:00": cannot parse " 02:00" as "Z07:00". Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=parsing time "2024-10-16T14:36:58 02:00" as "2006-01-02T15:04:05Z07:00": cannot parse " 02:00" as "Z07:00", metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://api.crc.testing:6443/api/v1/namespaces/redhat-ods-operator/pods/rhods-operator-76bb96fb9d-qg2cs/log?pretty=false&sinceTime=2024-10-16T14:36:58+02:00. Message: parsing time "2024-10-16T14:36:58 02:00" as "2006-01-02T15:04:05Z07:00": cannot parse " 02:00" as "Z07:00". Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=parsing time "2024-10-16T14:36:58 02:00" as "2006-01-02T15:04:05Z07:00": cannot parse " 02:00" as "Z07:00", metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:660)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:640)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:589)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleRaw(OperationSupport.java:739)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleRawGet(OperationSupport.java:474)
	at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.doGetLog(PodOperationsImpl.java:133)
	at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:141)
	at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:166)
	at io.fabric8.kubernetes.client.utils.internal.PodOperationUtil.getLog(PodOperationUtil.java:109)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.ReplicaSetOperationsImpl.getLog(ReplicaSetOperationsImpl.java:86)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.DeploymentOperationsImpl.getLog(DeploymentOperationsImpl.java:130)
	at io.fabric8.kubernetes.client.dsl.internal.apps.v1.RollableScalableResourceOperation.getLog(RollableScalableResourceOperation.java:89)
```

The problem appears to be that the `+` character in the timestamp is not urlencoded.

### Fabric8 Kubernetes Client version

6.13.0

### Steps to reproduce

```
client.pods().inNamespace("test").withName("foo").sinceTime("2024-10-16T14:36:58+02:00").getLog();
```

### Expected behavior

I'd expect fabric8 to urlencode the timestamp.

### Runtime

OpenShift

### Kubernetes API Server version

other (please specify in additional context)

### Environment

Linux, macOS

### Fabric8 Kubernetes Client Logs

_No response_

### Additional context

```
% oc version
Client Version: 4.17.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: 4.16.7
Kubernetes Version: v1.29.7+6abe8a1
```

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:testCompile (default-testCompile) on project kubernetes-test: Fatal error compiling: java.lang.NoSuchMethodError: 'void io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinitionSpec.<init>(java.util.List, io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceConversion, java.lang.String, io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinitionNames, java.lang.Boolean, java.lang.String, io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceSubresources, io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceValidation, java.lang.String, java.util.List)' -> [Help 1]
```

I'll figure it out, but yeah, feeling none the wiser than my younger self, dealing with stuff like that.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@
 #### Bugs
 * Fix #6038: Support for Gradle configuration cache
 * Fix #6214: Java generator does not recognize fields in CRDs other than metadata, spec, and status
+* Fix #6459: Pod log request sinceTime param correctly encoded
 * Fix #6632: Mock server creationTimestamp and deletionTimestamp formatted consistently (ISO 8601)
 
 #### Improvements

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/PodOperationContext.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/PodOperationContext.java
@@ -17,6 +17,7 @@ package io.fabric8.kubernetes.client.dsl.internal;
 
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.dsl.ExecListener;
+import io.fabric8.kubernetes.client.utils.URLUtils;
 import io.fabric8.kubernetes.client.utils.URLUtils.URLBuilder;
 import lombok.AllArgsConstructor;
 import lombok.Builder;
@@ -143,7 +144,8 @@ public class PodOperationContext {
     if (sinceSeconds != null) {
       sb.append("&sinceSeconds=").append(sinceSeconds);
     } else if (sinceTimestamp != null) {
-      sb.append("&sinceTime=").append(sinceTimestamp);
+      // https://github.com/fabric8io/kubernetes-client/issues/6459
+      sb.append("&sinceTime=").append(URLUtils.encodeToUTF(sinceTimestamp).replace("%3A", ":"));
     }
     if (tailingLines != null) {
       sb.append("&tailLines=").append(tailingLines);

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/behavior/LogTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/behavior/LogTest.java
@@ -1,0 +1,86 @@
+/*
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.behavior;
+
+import io.fabric8.kubernetes.api.model.PodBuilder;
+import io.fabric8.kubernetes.api.model.PodListBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.http.TestStandardHttpClient;
+import io.fabric8.kubernetes.client.http.TestStandardHttpClientFactory;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("Pod Log Behavior")
+class LogTest {
+
+  private TestStandardHttpClientFactory factory;
+  private KubernetesClient client;
+  private TestStandardHttpClient httpClient;
+
+  @BeforeEach
+  void setUp() {
+    factory = new TestStandardHttpClientFactory();
+    client = new KubernetesClientBuilder().withHttpClientFactory(factory).build();
+    httpClient = factory.getInstances().iterator().next();
+  }
+
+  @Nested
+  class WithReadyPod {
+
+    private String namespace = "default";
+    private String podName = "pod-get-log";
+
+    @BeforeEach
+    void setUp() {
+      namespace = "default";
+      podName = "with-ready-pod";
+      factory.expect("/api/v1/namespaces/" + namespace + "/pods", 200, client.getKubernetesSerialization().asJson(
+          new PodListBuilder()
+              .withNewMetadata().endMetadata()
+              .addToItems(new PodBuilder()
+                  .withNewMetadata().withName(podName).endMetadata()
+                  .withNewStatus().withPhase("Running").endStatus()
+                  .build())
+              .build()));
+    }
+
+    @Test
+    void getLog() {
+      factory.expect("/api/v1/namespaces/" + namespace + "/pods/" + podName + "/log", 200, "log contents");
+      final var result = client.pods().inNamespace(namespace).withName(podName).getLog();
+      assertThat(result).isEqualTo("log contents");
+    }
+
+    @Test
+    // https://github.com/fabric8io/kubernetes-client/issues/6459
+    void getLogSinceTimeEncodesParameter() {
+      factory.expect("/api/v1/namespaces/" + namespace + "/pods/" + podName + "/log", 200, "log contents");
+      final var result = client.pods().inNamespace(namespace).withName(podName).sinceTime("2024-10-16T14:36:58+02:00").getLog();
+      assertThat(result).isEqualTo("log contents");
+      assertThat(httpClient.getRecordedConsumeBytesDirects())
+          .last()
+          .extracting(r -> r.getRequest().uri().getRawQuery())
+          .asString()
+          .contains("sinceTime=2024-10-16T14:36:58%2B02:00");
+    }
+
+  }
+}

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/internal/PodOperationUtilTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/internal/PodOperationUtilTest.java
@@ -199,7 +199,7 @@ class PodOperationUtilTest {
   }
 
   @Test
-  void testGetLog() {
+  void getLog() {
     // Given
     PodResource p1 = mock(PodResource.class, Mockito.RETURNS_DEEP_STUBS);
     PodResource p2 = mock(PodResource.class, Mockito.RETURNS_DEEP_STUBS);

